### ISSUE TITLE
The script query on binary field returns incorrect results

### ISSUE THREAD

**Issue by dijingran** [elastic/elasticsearch#24785]
**Elasticsearch version**:  
5.2.2

**Plugins installed**: []

**JVM version** (`java -version`): 
 HotSpot 1.8.0_74

**OS version** (`uname -a` if on a Unix-like system):

Linux es.244.22.hadoop.sjz 2.6.32-573.22.1.el6.centos.plus.x86_64 #1 SMP Wed Mar 23 17:51:43 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

 1.  We used a binary field , and tried to use script query on that field, unfortunately  we got very weird results.
**es returns too much results, which script return false**. 

here is the mapping:

```
...
            "evtCnt": {
                  "properties": {
                     "code": {
                        "type": "binary",
                        "doc_values": true
                     }
                  }
               }
```
 2. setting:
```
            "refresh_interval": "-1",
            "number_of_shards": "16",
            "number_of_replicas": "0",
```
 3. source:
```
            "evtCnt": [
                  {
                     "code": [
                        "ATPGVQMEAQk=",
                        "ATPGVQMEAQk="
                     ]
                  }
               ]
```
4. myFilter.painless:
```java
boolean myFilter(List list, int fromDate, int toDate) {
                int totalCount = 0;

                for(int i=evt.size()-1;i>=0;i--){
                        byte[] bytes = list.get(i).bytes;
                        for(int j=0;j<bytes.length;){
                                ....
                                totalCount = totalCount + bytes[j++];
                                ....
                        }
                }
                return totalCount > 1;
    }
return myFilter(doc[params.column].values, params.from, params.to);
```
5. script query : 
```javascript
"query": {
      "bool": {
         "must": [
            {
                "exists":{
                    "field":"evtCnt.code"
                }
            },
            {
               "script": {
                  "script": {
                     "file": "myFilter",
                     "lang": "painless",
                     "params": {
                        "column": "evtCnt.code",
                        "from": 20170201,
                        "to": 20170515
                     }
                  }
               }
            }
         ]
      }
   }
```
6. The query result returns many wrong records. (which myFilter function returns false )

7. **At last we changed the mapping and source** (of cause a little change on script) , everything works fine.  

The mapping : 
```
             "evtCnt": {
                  "type": "binary",
                  "doc_values": true
               }
```
and changed the source to (single value, not an array) : 
```
           {
               "evtCnt": "ATPF7QEBAQEzxe8BAQQBM8Y9AQEB"
            }

```

**Comment by martijnvg** [comment]
@dijingran Thanks for reporting this issue. This looks like a bug. What happens here is that there is encoding logic that folds multiple binary array values into a single binary value, but at query time no decoding happens. Causing your `totalCount` var in your script to be always be 1 if the `evtCnt` is specified.

**Comment by martijnvg** [comment]
Turns out the decoding logic was not missing and behaving correctly. This bug was caused by something else: #25129

**Comment by martijnvg** [comment]
Turns out that #25129 only fixes a bug on master (6.0.0-alpha1 and 6.0.0-alpha2). This bug doesn't exist on 5.x because scrips access doc values in an array based way, whereas in master it is accessed in an iterator based manner which requires scripts to prefill an array. In that case #25129 is required.

@dijingran Can you share a more isolated example that demonstrates the issue that you're seeing?

**Comment by colings86** [comment]
No further feedback. If you are still able to reproduce this issue please provide details and we can re-open
